File: OC版本/EPJH/EPJH/Component/Root/RootNavigationController.m
//
//  RootNavigationController.m
//  EPJH
//
//  Created by Hans on 2020/7/15.
//  Copyright © 2020 hans3d. All rights reserved.
//

#import "RootNavigationController.h"

@interface RootNavigationController ()

@end

@implementation RootNavigationController

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view.
}

// 拦截所有push进来的控制器
- (void)pushViewController:(UIViewController *)viewController animated:(BOOL)animated{
    if (self.viewControllers.count > 0) {
   
        // 这时push进来的控制器viewController，不是第一个子控制器（不是根控制器）
        
        // viewController.hidesBottomBarWhenPushed = YES;
        UIButton *btn = [UIButton buttonWithType:UIButtonTypeCustom];
        btn.frame = CGRectMake(0, 0, 30, 50);
        [btn addTarget:self  action:@selector(backAction) forControlEvents:UIControlEventTouchUpInside];
        [btn setImage:[UIImage imageNamed:@"icon_common_back"] forState:UIControlStateNormal];
        viewController.navigationItem.leftBarButtonItem = [[UIBarButtonItem alloc] initWithCustomView:btn];
    
    }
    
    [super pushViewController:viewController animated:animated];
}

- (void)backAction{
    [self popViewControllerAnimated:YES];
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

/*
 #pragma mark - Navigation
 
 // In a storyboard-based application, you will often want to do a little preparation before navigation
 - (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
 // Get the new view controller using [segue destinationViewController].
 // Pass the selected object to the new view controller.
 }
 */

@end

File: Swift版本/EPJH/EPJH/Component/Root/RootNavigationController.swift
//
//  RootNavigationController.swift
//  EPJH
//
//  Created by Hans on 2020/8/3.
//  Copyright © 2020 hans3d. All rights reserved.
//

import UIKit

class RootNavigationController: UINavigationController {

    override func viewDidLoad() {
        super.viewDidLoad()
        
    }
    
    /**重写push方法*/
    override func pushViewController(_ viewController: UIViewController, animated: Bool) {
        
        self.edgesForExtendedLayout = UIRectEdge(rawValue: 0)
        
        if self.viewControllers.count > 0 {
            
            viewController.hidesBottomBarWhenPushed = true
            
            let  leftBtn = UIButton.init(type: .custom)
            leftBtn.bounds = CGRect.init(x: 0, y: 0, width: 40, height: 18)
            leftBtn.addTarget(self, action: #selector(RootNavigationController.popViewController(animated:)), for: .touchUpInside)
            leftBtn.setImage(UIImage.init(named: "icon_common_back"), for: UIControl.State())
            leftBtn.imageEdgeInsets = UIEdgeInsets(top: 0, left: -12, bottom: 0, right: 12)
            let leftItem = UIBarButtonItem.init(customView: leftBtn)
            viewController.navigationItem.leftBarButtonItem = leftItem
      
        }
        
        super.pushViewController(viewController, animated: animated)
    }
    
    
    func back() {
        popViewController(animated: true)
    }
    
    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    
    
    /*
     // MARK: - Navigation
     
     // In a storyboard-based application, you will often want to do a little preparation before navigation
     override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
     // Get the new view controller using segue.destinationViewController.
     // Pass the selected object to the new view controller.
     }
     */
    

}
